lambda/fulfillment/lib/middleware/2_preprocess.js

```diff
@@ -60,53 +64,49 @@ async function update_userInfo(userId, req_userInfo) {
-        var result = redacted_text != text;
-        console.log(`Is Redacted ${result}`)
-        if(result) //if the regex was returned. No need to call Comprehend
-            return result;
+    qnabot.log("Testing redaction ")
+    let found_redacted_pii = false
+    if (piiRegex) {
+        let re = new RegExp(piiRegex, "g");
+        let redacted_text = text.replace(re, "XXXXXX");
+        found_redacted_pii = redacted_text != text;
     } else {
-        console.log("Warning: No value found for setting  PII_REJECTION_REGEX not using REGEX Matching")
+        qnabot.log("Warning: No value found for setting  PII_REJECTION_REGEX not using REGEX Matching")
     }
-    if(useComprehendForPII){
+    if (useComprehendForPII) {
         var params = {
-                LanguageCode: "en",
-                Text: text
-            };
-            try
-            {
-                var comprehendResult = await comprehend_client.detectPiiEntities(params).promise();
-                console.log(JSON.stringify(comprehendResult) + "entity count == " + comprehendResult.Entities.length )
-                if(!("Entities" in comprehendResult) ||  comprehendResult.Entities.length == 0)
-                {
-                    console.log("No PII found by Comprehend")
-                    return false;
-                }
-                console.log("Ignoring types for PII == " + pii_rejection_ignore_list)
-                pii_rejection_ignore_list = pii_rejection_ignore_list.toLowerCase().split(",")
-
-                return comprehendResult.Entities.filter(entity => entity.Score > 0.90 && pii_rejection_ignore_list.indexOf(entity.Type.toLowerCase()) == -1).length > 0;;
-
-            }catch(exception)
-            {
-                console.log("Warning: Exception while trying to detect PII with Comprehend. Skipping...");
-                console.log("Exception " + exception);
+            LanguageCode: "en",
+            Text: text
+        };
+        try {
+            var comprehendResult = await comprehend_client.detectPiiEntities(params).promise();
+            process.env.comprenhend_pii = comprehendResult //This will be used by the filter.
+            if (!("Entities" in comprehendResult) || comprehendResult.Entities.length == 0) {
+                qnabot.log("No PII found by Comprehend")
                 return false;
             }
-    
+            qnabot.log("Ignoring types for PII == " + pii_rejection_ignore_list)
+            pii_rejection_ignore_list = pii_rejection_ignore_list.toLowerCase().split(",")
+            let entitiesToFilter = comprehendResult.Entities.filter(entity => entity.Score > 0.90 && pii_rejection_ignore_list.indexOf(entity.Type.toLowerCase()) == -1)
+            //For now, we *redact* all detected PII from CloudWatch.  We accept any PII for processing that is listed PII_REJECTION_IGNORE_TYPES
+
+            return entitiesToFilter.length > 0 || found_redacted_pii;;
```

Comment:
Earlier we were checking found_redacted_pii in the piiRegex if statement above and returning when true. So, if we would not use the comprehend detectPiiEntities in that case as we already detected PII.

However, this behavior is now changed to use comprehend regardless of the piiRegex detecting PII. Is this intentional change? It results in extra calls to comprehend in that case.